test/es6/iteratorclose.js

```diff
@@ -0,0 +1,982 @@
+	},    
+    {
+		name : "For..of - .return function can also throw",
+		body : function () {
+            shouldReturnThrow = true;
+            assert.throws(function () { for (var i of iterable) { break; } }, Error, ".return function throws", "Exception from return function");
+            shouldReturnThrow = false;
+		}
+	},    
+    {
+		name : "For..of - caller and .return function both throw and caller's exception wins",
+		body : function () {
+            shouldReturnThrow = true;
+            returnCount = 0;
+            assert.throws(function () { for (obj.prop of iterable) { break; } }, Error, "Setting value will throw", "From setter");
+            assert.areEqual(returnCount, 1, "Ensuring that abrupt loop completion due to exception will call return function");
+            shouldReturnThrow = false;
+		}
+	},    
+    {
+		name : "For..of - with generator",
+		body : function () {
+            var finallyCount = 0;
+            function *gf() {
+                try {
+                    yield 1;
+                    assert.fail('Should not reach here after yield');
+                }
+                finally {
+                    finallyCount++;
+                }
+                assert.fail('Should not reach here after finally');;
+            }
+            
+            for (var i of gf()) {
+                break;
+            }
+
+            assert.areEqual(finallyCount, 1, "'break' causes the return function called which should execute the finally block");
+            
+            finallyCount = 0;
+            assert.throws(function () { for (obj.prop of gf()) { } }, Error, "Setting value will throw", "From setter");
+            assert.areEqual(finallyCount, 1, "Exception causes the return function called which should execute the finally block");
+            
+            function* gf2() {
+                yield 1;
+                assert.fail('Should not reach here after yield');
+            }
+            var returnCount = 0;
+            gf2.prototype.return = function() {
+                returnCount++;
+                return {};
+            };
+            
+            for (var i of gf2()) {
+                break;
+            }
+
+            assert.areEqual(returnCount, 1, "Loop break due to 'break' should call the return function on the generator");
+            
+            gf2.prototype.return = function() {
+                returnCount++;
+                throw new Error('Exception from return function');
+            };
+            assert.throws(function () { for (i of gf2()) { break; } }, Error, "return function throws", "Exception from return function");
+            returnCount = 0;
+            assert.throws(function () { for (obj.prop of gf2()) { } }, Error, "Exception at destructuring element wins", "From setter");
+            assert.areEqual(returnCount, 1, "Exception in loop should call the return function");
+            
+		}
+	},    
+    {
+		name : "Iterator close with yield *",

```

Comment:
This test is actually verifying that yield* calls the inner generator with the same api call (next, throw, or return) that was used to invoke itself.  This is behavior we already had prior to your change and I believe we have tests for in the generator test suite.  These tests here are redundant.

Instead you'll need to verify that a yield* expression itself calls return() upon a throwing inner iterable.  E.g. something like

```javascript
function* test() {
    try {
        yield* iterableWithReturnMethodThatThrowsInNextMethod;
    } catch (e) {
    }
    // verify here that the return method was called
}
var g = test();
g.next();
g.next();
g.next(); // however many times to make the throw occur
```